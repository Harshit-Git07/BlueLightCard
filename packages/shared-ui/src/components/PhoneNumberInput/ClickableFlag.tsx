import { FlagImage, ParsedCountry } from 'react-international-phone';
import { FC } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCaretDown, faCaretUp } from '@fortawesome/pro-solid-svg-icons';

export type Props = {
  country: ParsedCountry;
  isOpen: boolean;
  toggleDropdown: () => void;
  disabled?: boolean;
  className: string;
  isSelectable?: boolean;
  controlsId?: string;
};

const ClickableFlag: FC<Props> = ({
  country,
  isOpen,
  toggleDropdown,
  disabled,
  className,
  isSelectable,
  controlsId,
}) => {
  return (
    <button
      title={country.name}
      aria-label="Country selector"
      aria-haspopup="listbox"
      aria-expanded={isOpen}
      aria-controls={controlsId}
      type="button"
      disabled={disabled}
      data-country={country.iso2}
      className={`flex flex-row items-center ${!isSelectable ? 'cursor-default' : 'cursor-pointer'} ${className}`}
      onClick={isSelectable ? toggleDropdown : undefined}
    >
      <FlagImage iso2={country.iso2} className={`h-full w-8 rounded-sm`} />
      {isSelectable && (
        <FontAwesomeIcon
          className="h-3 w-3 text-colour-onSurface-light dark:text-colour-onSurface-dark p-1"
          size="2xs"
          icon={isOpen ? faCaretUp : faCaretDown}
        />
      )}
    </button>
  );
};

export default ClickableFlag;
