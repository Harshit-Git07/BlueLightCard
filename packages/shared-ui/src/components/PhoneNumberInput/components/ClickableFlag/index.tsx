import { FlagImage } from 'react-international-phone';
import { FC } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCaretDown, faCaretUp } from '@fortawesome/pro-solid-svg-icons';

export type Props = {
  isOpen: boolean;
  toggleDropdown: () => void;
  disabled?: boolean;
  className: string;
  isSelectable?: boolean;
  controlsId?: string;
  name?: string;
  iso2?: string;
};

const ClickableFlag: FC<Props> = ({
  isOpen,
  toggleDropdown,
  disabled,
  className,
  isSelectable,
  controlsId,
  iso2 = '',
  name = '',
}) => {
  return (
    <button
      title={name}
      aria-label="Country selector"
      aria-haspopup="listbox"
      aria-expanded={isOpen}
      aria-controls={controlsId}
      type="button"
      disabled={disabled}
      data-country={iso2}
      className={`flex flex-row items-center ${!isSelectable ? 'cursor-default' : 'cursor-pointer'} ${className}`}
      onClick={isSelectable ? toggleDropdown : undefined}
    >
      <FlagImage iso2={iso2} className={`h-full w-[32px] rounded-sm`} />
      {isSelectable && (
        <FontAwesomeIcon
          className="h-3 w-3 text-colour-onSurface-light dark:text-colour-onSurface-dark p-1"
          size="2xs"
          icon={isOpen ? faCaretUp : faCaretDown}
        />
      )}
    </button>
  );
};

export default ClickableFlag;
