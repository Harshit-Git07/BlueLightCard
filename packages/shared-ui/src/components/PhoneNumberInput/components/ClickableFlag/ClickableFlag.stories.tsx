import { Meta, StoryFn } from '@storybook/react';
import ClickableFlag from '.';

const componentMeta: Meta<typeof ClickableFlag> = {
  title: 'Component System/PhoneNumberInput/ClickableFlag',
  component: ClickableFlag,
};

const DefaultTemplate: StoryFn<typeof ClickableFlag> = (args) => <ClickableFlag {...args} />;

export const Default = DefaultTemplate.bind({});

Default.args = {
  iso2: 'gb',
  name: 'UK',
  isOpen: false,
  toggleDropdown: () => undefined,
};

export const IsOpen = DefaultTemplate.bind({});

IsOpen.args = {
  iso2: 'gb',
  name: 'UK',
  isOpen: true,
  toggleDropdown: () => undefined,
};

export default componentMeta;
