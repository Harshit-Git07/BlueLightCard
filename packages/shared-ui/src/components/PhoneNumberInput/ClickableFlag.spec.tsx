import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ClickableFlag, { Props } from './ClickableFlag';
import { defaultCountries, parseCountry } from 'react-international-phone';

const ukCountry = defaultCountries.map(parseCountry).find((country) => country.iso2 === 'gb')!;

describe('ClickableFlag component', () => {
  const mockToggleDropdown = jest.fn();
  const defaultProps: Props = {
    country: ukCountry,
    isOpen: false,
    toggleDropdown: mockToggleDropdown,
    disabled: false,
    className: '',
    isSelectable: true,
    controlsId: 'countryDropdown',
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render component without error', () => {
    const { baseElement } = render(<ClickableFlag {...defaultProps} />);
    expect(baseElement).toBeTruthy();
  });

  it('matches open and closed snapshots', () => {
    const { container: closedContainer } = render(
      <ClickableFlag {...defaultProps} isOpen={false} />,
    );
    expect(closedContainer).toMatchSnapshot('ClickableFlag - closed state');

    const { container: openContainer } = render(<ClickableFlag {...defaultProps} isOpen={true} />);
    expect(openContainer).toMatchSnapshot('ClickableFlag - open state');
  });

  it('triggers callback when clicked and isSelectable is true', () => {
    render(<ClickableFlag {...defaultProps} isSelectable={true} />);
    screen.getByRole('button').click();
    expect(mockToggleDropdown).toHaveBeenCalled();
  });

  it('should be disabled when the disabled prop is true', () => {
    render(<ClickableFlag {...defaultProps} disabled={true} />);
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
  });

  it('does not trigger callback when isSelectable is false', () => {
    render(<ClickableFlag {...defaultProps} isSelectable={false} />);
    screen.getByRole('button').click();
    expect(mockToggleDropdown).not.toHaveBeenCalled();
  });

  it('applies custom className correctly', () => {
    const customClassName = 'custom-class';
    render(<ClickableFlag {...defaultProps} className={customClassName} />);
    const button = screen.getByRole('button');
    expect(button).toHaveClass(customClassName);
  });

  it('has correct accessibility attributes', () => {
    render(<ClickableFlag {...defaultProps} />);
    const button = screen.getByRole('button');

    expect(button).toHaveAttribute('aria-controls', 'countryDropdown');
    expect(button).toHaveAttribute('aria-expanded', 'false');
  });
});
