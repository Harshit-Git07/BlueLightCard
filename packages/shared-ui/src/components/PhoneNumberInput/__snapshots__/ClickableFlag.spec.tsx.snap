// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickableFlag component matches open and closed snapshots: ClickableFlag - closed state 1`] = `
<div>
  <button
    aria-controls="countryDropdown"
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-label="Country selector"
    class="flex flex-row items-center cursor-pointer "
    data-country="gb"
    title="United Kingdom"
    type="button"
  >
    <img
      alt=""
      class="react-international-phone-flag-emoji h-full w-8 rounded-sm"
      data-country="gb"
      draggable="false"
      loading="lazy"
      src="https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/svg/1f1ec-1f1e7.svg"
    />
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-caret-down fa-2xs h-3 w-3 text-colour-onSurface-light dark:text-colour-onSurface-dark p-1"
      data-icon="caret-down"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M137.4 374.6c12.5 12.5 32.8 12.5 45.3 0l128-128c9.2-9.2 11.9-22.9 6.9-34.9s-16.6-19.8-29.6-19.8L32 192c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9l128 128z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`ClickableFlag component matches open and closed snapshots: ClickableFlag - open state 1`] = `
<div>
  <button
    aria-controls="countryDropdown"
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-label="Country selector"
    class="flex flex-row items-center cursor-pointer "
    data-country="gb"
    title="United Kingdom"
    type="button"
  >
    <img
      alt=""
      class="react-international-phone-flag-emoji h-full w-8 rounded-sm"
      data-country="gb"
      draggable="false"
      loading="lazy"
      src="https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/svg/1f1ec-1f1e7.svg"
    />
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-caret-up fa-2xs h-3 w-3 text-colour-onSurface-light dark:text-colour-onSurface-dark p-1"
      data-icon="caret-up"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M182.6 137.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8H288c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
